parser.go

```diff
@@ -3796,3 +3931,31 @@ func (p *Parser) errorfAtToken(tok *token.Token, msg string, params ...interface
 func (p *Parser) panicfAtToken(tok *token.Token, msg string, params ...interface{}) {
 	panic(p.errorfAtToken(tok, msg, params...))
 }
+
+// This function can't be a method because Go haven't yet supported generic methods.
+func parseSeparatedList[T interface {
+	ast.Node
+	comparable
+}](p *Parser, sep token.TokenKind, doParse func() T) []T {
```

Comment:
I have some concerns about this function.

- When `doParse` returns the zero value, it terminates the iteration. If the parser state is updated in `doParse`, the behavior is not intuitive.
- Is there any situation to pass a value other than `","` to `sep` in this function? Or, maybe we want to pass multiple separator tokens or more generally a function to parse separators.
- "Trailing comma" is frequently used in such situations, but is it not addressed?
- Given these issues, the function can be complex and unwieldy.
- That is, I believe that syntax separated by something is too general for parsing and should not be abstracted.

I agree with `tryExpect`. I also think that adding a function like `parseIdentList` instead of this function would not be a bad idea.

Reply:
> When doParse returns the zero value, it terminates the iteration. If the parser state is updated in doParse, the behavior is not intuitive.

It may mean that treating `parse*` and `tryParse*` in the same function can be confusing.

> Is there any situation to pass a value other than "," to sep in this function? Or, maybe we want to pass multiple separator tokens or more generally a function to parse separators.

> "Trailing comma" is frequently used in such situations, but is it not addressed?

 Currently, it seems that trailing comma is [only permitted in SELECT list](https://cloud.google.com/spanner/docs/reference/standard-sql/lexical#trailing_commas) in GoogleSQL.
So, I have not yet considered to generalize it.

> Is there any situation to pass a value other than "," to sep in this function? Or, maybe we want to pass multiple separator tokens or more generally a function to parse separators.

I have considered `NEXT` of GQL.
https://cloud.google.com/spanner/docs/reference/standard-sql/graph-query-statements#gql_syntax

```
multi_linear_query_statement:
  linear_query_statement
  [
    NEXT
    linear_query_statement
  ]
  [...]
```

(It may be needed to handle `expect` and `expectKeywordLike`)
Token sequence or more general parser function as separator may be appeared, but they will be much more complex.(e.g. need to know `UNION ALL ` or `UNION DISTINCT`)

Comma separated with not `tryParse*` may be the most frequently used pattern.
Other usecases in `CREATE PROPERTY GRAPH` repo. https://github.com/cloudspannerecosystem/memefish/pull/101/files#diff-83eb8e32639d01cf443d6d8bde24c1c8be78766090d8c5f8586c36250cfedca6

```
	elements := parseSeparatedList(p, ",", p.parsePropertyGraphElement)
```

```
	list := parseSeparatedList(p, ",", func() *ast.PropertyGraphDerivedProperty {
```

Reply:
I will drop the condition of `doParse() == zero`, and research other usecases in the codebase.

Reply:
すみません。自分のこだわり的な側面もかなりつよいので諦めて下さい。自分の目が黒いうちはこの関数がコードに入ることはないです

Reply:
OK. I will remove this abstracted function.

Instead, it is better to have the reference implementation of comma separated list to avoid bugs and code skews.

Reply:
Typically, parsing a comma-separated list is done by the following pattern:

```go
	xxxs := []*ast.XXX{p.parseXXX()}
	for p.Token.Kind == "," {
		p.nextToken()
		xxxs = append(xxxs, p.parseXXX())
	}
```

I quickly checked how much this idiom is used by `grep 'for p.Token.Kind == ","' parser.go | wc -l`, and it answered `19`. Therefore, we could introduce `parseCommaSepList` for that purpose. The reason I did not implement this function in the past is that Go did not have generics, so the code would not be short at all.

However, this function should always be used on a single line. That is, this usage seems wrong.

https://github.com/cloudspannerecosystem/memefish/blob/fdefffb4229337749296a510b8b525fe389e52ae/parser.go#L2664-L2672

Reply:
`parseCommaSepList(p, p.parseSomething)` will be very helpful. Thanks!

Reply:
https://github.com/cloudspannerecosystem/memefish/pull/141

Reply:
Already migrated https://github.com/cloudspannerecosystem/memefish/pull/111/commits/a90745cba4b9eb574f5e1383688d7f69ea865c45